be/src/formats/parquet/page_reader.cpp

```diff
@@ -47,18 +50,40 @@ Status PageReader::next_header() {
     size_t remaining = _finish_offset - _offset;
     uint32_t header_length = 0;
 
+    RETURN_IF_ERROR(_stream->seek(_offset));
+
     do {
         allowed_page_size = std::min(std::min(allowed_page_size, remaining), kMaxPageHeaderSize);
 
         std::vector<uint8_t> page_buffer;
-        page_buffer.reserve(allowed_page_size);
-        uint8_t* page_buf = page_buffer.data();
-
-        RETURN_IF_ERROR(_stream->read_at_fully(_offset, page_buf, allowed_page_size));
+        const uint8_t* page_buf = nullptr;
+
+        // prefer peek data instead to read data.
+        bool peek_mode = false;
+        {
+            auto st = _stream->peek(allowed_page_size);
+            if (st.ok() && st.value().size() == allowed_page_size) {
+                page_buf = (const uint8_t*)st.value().data();
+                peek_mode = true;
+            } else {
+                page_buffer.reserve(allowed_page_size);
+                RETURN_IF_ERROR(_stream->read_at_fully(_offset, page_buffer.data(), allowed_page_size));
+                page_buf = page_buffer.data();
+                auto st = _stream->peek(allowed_page_size);
+                if (st.ok()) {
+                    _stats->bytes_read -= allowed_page_size;
+                    peek_mode = true;
+                }
+            }
+        }
 
         header_length = allowed_page_size;
         auto st = deserialize_thrift_msg(page_buf, &header_length, TProtocolType::COMPACT, &_cur_header);
+
         if (st.ok()) {
+            if (peek_mode) {
+                _stats->bytes_read += header_length;
```

Comment:
why not count when peek_mode = false?

Reply:
> why not count when peek_mode = false?

I guess because the bytes_read is counted in`read_at_fully` function when peek_mode=false

Reply:
Yes. If not peek mode, it means we have already taken into bytes read.

For example, if there is shared buffer underneath, then `read_at_fully` will trigger app & fs layers to read.